### PR TITLE
minor improvements

### PR DESCRIPTION
**Pull request by tessus**
- added a Makefile
- fixed Xcode warnings and deprecations
- minor changes to README.md

The build phase to copy the plugin to `~/Library/Quicklook` does not work. Xcode complains about the absolute path `/~/Library/Quicklook`, which of course cannot work. Therefore I created a build script `Makefile` to copy the plugin instead. -> `make && make install`.

Also, you might not want to install and replace your installed version every time you build the product.

### CODE REVIEW

**Comment by tessus** [comment]
@tsdorsey have you looked at my PR? what do you think?

**Comment by tsdorsey** [comment]
@tessus I have, this looks great. Thank you! I'm slammed until April but I will test and merge this when I can. I don't even have Xcode installed right now :)

**Comment by tessus** [comment]
No worries, all good. I just thought that you missed the notification.

**Comment by ianleon** [comment]
You can replace the ~ with $HOME.
It avoids needing a workaround like a makefile and it works great

**Comment by tessus** [comment]
@ianleon I added the Makefile because I like to build on the command line and not use the UI. Also, there's a difference between compiling and installing. The current build phase does both.

**Comment by ianleon** [comment]
I like to use the command line too. I use makefiles a lot. But by doing this you are introducing something that not everyone is familiar with. My solution gets around that absolute path issue without introducing a makefile.

**Comment by ianleon** [comment]
Checkout my fork https://github.com/ianleon/qlstephen

**Comment by ianleon** [comment]
Basically my idea with it was to fix the warnings, fix that installation error, and not introduce much new stuff.

**Comment by tessus** [comment]
> But by doing this you are introducing something that not everyone is familiar with. 

I'm sorry, but this makes no sense. People who know what a Makefile is will probably use it. People who don't will ignore it and build the same way as before.

> My solution gets around that absolute path issue without introducing a makefile.

Ok. So how do you compile it without installing it? Auomatically installing during build is baad practice. You should never, ever do that.

Don't worry, it's not gonna be merged. @tsdorsey had almost a year to merge the change. Nobody can tell me that 5 minutes are too much time to invest.

**Comment by ianleon** [comment]
> People who don't will ignore it and build the same way as before.

But as far as I can see you didn't fix THIS issue. you only fixed it for people who know what a makefile is, and know how to use it. 

You could have FIXED the actual issue AND added the makefile for people who know what to that is. 

And I don't know if its such a bad practice to do that. The original authors did it and Xcode supports it.

**Comment by tessus** [comment]
> You could have FIXED the actual issue AND added the makefile for people who know what that is.

No, because then a `make` will also install the file.

> And I don't know if its such a bad practice to do that. The original authors did it and Xcode supports it.

Yes, it is bad practice. Have you ever read books about software engineering and deployment? Just because the original authors did something, doesn't mean it was a good idea.

Anyway, this project is dead. So nothing's gonna happen.

**Comment by ianleon** [comment]
Correct me if i'm wrong but you removed the ability for the project to actually install the end product on the user's QuickLook library.

**Comment by ianleon** [comment]
OK what book is this ?

**Comment by tessus** [comment]
I'm done here. I suggest that you open your own PR and stop hijacking mine.

> OK what book is this ?

Seriously? Any book that profs use at university to teach software design principles. That was 94 for me, so I don't know the names of current books. Or when you look at Docker and Kubernetes deployment handbooks. Google it for crying out loud, why compiling and installing SW in one step is a bad idea. (Or at least it's a no-go without the option to not install it. `make install` will compile and install in seemingly one step, but in that case the person who uses it knows what they are doing.)

**Comment by ianleon** [comment]
Most people want to build and install this project on their own because simply installing it through brew / port as they used to has broken down because of the latest changes in the OS.  Its a little easier to open the Xcode project and HIT PLAY. 

You want to destroy that simplicity with your make; make install; UNIX PURISM and your "GO READ A BOOK" insults.

Thats not what its about in this environment. You might be getting confused with compiling C / C++

**Comment by tessus** [comment]
I try it one more time.

My argument has nothing to do with unix purism or any insults.

No software project should install a program when on clicks the `Build` button. The build button is for building a product, not for installing it. 

- You can go to `Product -> Archive` and create a file you can double-click to install.
- It's also ok to create a separate schema which includes that step, but you can't do it by default. 
- Or you go to the directory where the target file is and double-click it.
- Or you right click the file in `Product Navigator -> Products -> qlStephen` and say open in Finder and then double click to install it.

Imagine a SW project you compile and all of a sudden it overwrites the executable on the production system. You can't really think that this is the way to go. If you do, you probably won't have a job for very long.
Nobody would compile on a production system you say. I'd hope so too, but I have seen this insanity.
I also don't want to install a product on my build machine, which is only there for building. A compiled product should be shipped to the test and/or integration system to run through FVT and SVT. 

This environment still should have a sane default. And installing without permission is not sane.

**Comment by ianleon** [comment]
> Imagine a SW project you compile and all of a sudden it overwrites the executable on the production system. You can't really think that this is the way to go. If you do, you probably won't have a job for very long.

Of course you wouldn't want to do that when building something like a web service. I do work as a web developer and im pretty good at that kind of stuff. I use makefiles HEAVILY. I use git for deployment, ssh keys, I have makefiles that set up staging environments for me ..etc 

Im also familiar with Linux servers and the process for building from source ./configure; make; make install. 

I don't see how all of that rigidity is required HERE. 

I feel that your solution which only applies to people using the command line is wrong and exclusive of people who want to use the GUI. You have literally removed something that was put in there on purpose to make things easier. People who don't use the command line and don't know what make is don't matter to you, that's not very nice.

> No software project should install a program when one clicks the Build button. The build button is for building a product, not for installing it.

It happens to be the case that the project CAN be set up to work this way. 

And I did try what you said about double clicking the product, and making the archive and it did not work. And I did test it with running qlmanage -r before and after. The only thing that worked was dragging the generated product to the appropriate location. Why should someone have to either be forced to use the command line OR copy the file to the library using finder. Thats too complicated and finicky. You broke something that was easy and correct and made it hard and complex because it makes sense to YOU and way of doing things. Maybe you should seek to understand first before you go on deleting stuff that other people like because you deem them to be wrong from YOUR point of view. 

In conclusion, this environment is different, it may not be what you are used to. Most your concerns are actually taken care of, just probably in a way that you probably don't comprehend just yet because you are comparing it to something else that's more familiar to you.

**Comment by ianleon** [comment]
Also, this is not about people compiling this tiny thing on their BUILD computer and then moving it to their REGULAR USE computer. People just need a tool to quickly look at text files on their laptops. Its simple, the faster you make it for people the better. Focus on contributing fixes to actual problems. Not changing how things are done.

**Comment by tessus** [comment]
You still haven't explained to me how I can compile it without installing it.

Don't forget that the current copy step does not work anyway (see #85), because of the error I mentioned earlier. So I'm not breaking anything.

**Comment by ianleon** [comment]
You can compile it without installing by removing the Copy Files Build Phase. Which is basically what you did. And then you added the Makefile. Which only works from the command line.

The current Copy Files Build Phase was broken, I fixed it by replacing ~ with $HOME. Which again is FIXING a problem.

**Comment by ianleon** [comment]
If someone doesn't go through the trouble of notarizing the app with their Developer ID, then the only way to install this thing would be by actually building it from source. And if that's going to be the only way, shouldn't we make it as simple as possible for people?

**Comment by tessus** [comment]
As I said, open your PR and that's that. I just cleaned up the mess the original author made.
You apparrently want to fix the mess by replacing the not working mess with a working mess. That's fine. Open your PR. You are entitled to your opinion. I'm entitled to mine.

>  shouldn't we make it as simple as possible for people?

Yes, but your approach won't work either for people who compile it themselves. They have to do a lot more - at least on the latest macOS.
And also, if someone can clone a repo, they most likely can do a `make install` as well. 

It's the developer's decision anyway, whether my PR gets merged or not. And this project is pretty much dead.

So please open your PR and maybe you could just add another schema that includes the install step. You can then add a sentence to the README how to build and install at the same time.
I'm not inherently against it. it just shouldn't be the default.

**Comment by ianleon** [comment]
I did not make a PR, but I made a fork. I think your PR is almost correct, but not quite. that's why im giving feedback. I strongly feel that it should not be implemented this way. Thats what THIS CONVERSATION section is all about.

**Comment by ianleon** [comment]
You can't be telling people to go away like this. Thats not nice.

**Comment by tessus** [comment]
We won't come to an agreement here. I don't think that it should be installed when doing a build. That's my firm believe and I won't be convinced otherwise.

I told you that you could create your own PR.

I will never add anything to my PR that will install the product automatically after the build process - at least not by default. No matter what you say, it won't change my opinion.

**Comment by ianleon** [comment]
I’m on the latest os and this method worked for me that’s the whole reason why I published a fork. 

Part of me thinks like you. I would like to forget everything else and use make everywhere. 
Make is awesome and I get it. But it’s wrong here. 

I’m on Catalina and this method works.

There’s no point in only building this tool. What are you going to do ? Open it in a hex reader?

**Comment by tessus** [comment]
> There’s no point in only building this tool. What are you going to do ? Open it in a hex reader?

Funny. No, but when I started to write code in this project (and this change was not the only thing I changed) and built the code to debug it, all of sudden my working copy would have been overwritten with a broken version. I was happy that the step was actually broken, but I really thought: what if it had not been?

For development the system should not be modified. Some things you can even debug within Xcode w/o having to install it on the system - if not, VMs are your friend.

This is a repo for development. If you want to make it easy for people, create a release tarball with an install .pkg.

**Comment by ianleon** [comment]
VMs really ?

**Comment by ianleon** [comment]
Also what’s the big deal if you break this little thing on your dev machine. It’s probably a good thing that you have to should eat your own dog food.

**Comment by ianleon** [comment]
“Eating your own dog food” now THAT is a good piece of software engineering wisdom.

**Comment by ianleon** [comment]
https://en.m.wikipedia.org/wiki/Eating_your_own_dog_food

**Comment by ianleon** [comment]
https://www.parkersoftware.com/blog/are-you-eating-your-own-dog-food/

**Comment by ianleon** [comment]
https://www.zdnet.com/article/microsoft-eat-your-own-dog-food/

**Comment by ianleon** [comment]
https://blog.codeenigma.com/eating-your-own-dog-food-17b348aada1a

**Comment by ianleon** [comment]
https://www.quora.com/What-does-dogfooding-mean-in-the-context-of-software-development

**Comment by ianleon** [comment]
Read it
